### PR TITLE
app/internal/window: [wasm] add support for insets and cache window-size

### PR DESCRIPTION
**Pull request by Inkeliz**
Currently, on iOS, the keyboard can block the content since it doesn't
trigger the "resize". Now, the `insets` will provide the size of the
keyboard (on iOS) and scrollbars. It's compatible with the Android (and
desktop), but Chrome automatic resizes the windows, so the `Inset` will
be 0.

The `Visual Viewport API` is also triggered when zooming, so the scale 
will change.

### CODE REVIEW

File: app/internal/window/os_js.go [hunk 1]
```diff
@@ -123,10 +129,19 @@ func (w *window) cleanup() {
...
+			w.draw(true)
+			return nil
+		})
+	} else {
+		w.addEventListener(w.visualViewport, "resize", func(this js.Value, args []js.Value) interface{} {
+			w.resize()
+			w.draw(true)
+			return nil
+		})
+	}
```
Comment (eliasnaur) [inline app/internal/window/os_js.go:144]: Deduplicate the if and else branches by introducing a local:

```
viewport := w.visualViewport
if viewport.IsUndefined() {
    viewport = w.window
}
```

File: app/internal/window/os_js.go [hunk 2]
```diff
@@ -466,43 +481,61 @@ func (w *window) ShowTextInput(show bool) {
...
-	if cfg == (unit.Metric{}) || width == 0 || height == 0 {
+	width, height, insets, metric := w.config()
+	if metric == (unit.Metric{}) || width == 0 || height == 0 {
 		return
 	}
-	w.mu.Lock()
-	w.scale = float32(scale)
-	w.mu.Unlock()
+
+	if w.resized {
```
Comment (eliasnaur) [inline app/internal/window/os_js.go:508]: Why delay resizing the canvas, and not do it immediately in w.resize?

**Review by eliasnaur — CHANGES_REQUESTED** [review]
Nice work.

File: app/internal/window/os_js.go [hunk 3]
```diff
@@ -466,43 +475,59 @@ func (w *window) ShowTextInput(show bool) {
...
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	w.scale = float32(w.window.Get("devicePixelRatio").Float())
+
+	rect := w.cnv.Call("getBoundingClientRect")
+	w.size.X = float32(rect.Get("width").Float()) * w.scale
+	w.size.Y = float32(rect.Get("height").Float()) * w.scale
+
+	if vx, vy := w.visualViewport.Get("width"), w.visualViewport.Get("height"); !vx.IsUndefined() && !vy.IsUndefined() {
```
Comment (eliasnaur) [inline app/internal/window/os_js.go:488]: Now that w.visualViewport is always set, is the IsUndefined tests still necessary?

Comment (Inkeliz) [reply app/internal/window/os_js.go:488]: Yes. The `w.visualViewport` can be either "Window" or "Visual Viewport". The `w.window` doesn't have `window.width`/`window.height`.

**Comment by eliasnaur** [comment]
Thanks! Merged.

**Pull request closed without merging**